### ISSUE TITLE
Memory leak when reusing variables with slim

### ISSUE THREAD

**Issue by koelscha** [tensorflow/tensorflow#12704]
### System information:
- Windows 7 x64
- Python 3.5.2 |Anaconda 4.2.0 (64-bit)
- Tensorflow 1.3.0 installed via pip

### Problem
I want to use TF-Slim models to classify images in a server. For this, I would like to load the network only once and reuse the variables. I adjusted the tutorial given here: https://github.com/tensorflow/models/blob/master/slim/slim_walkthrough.ipynb

### Code to reproduce
```
import tensorflow as tf
from tensorflow.contrib import slim

import urllib.request as urllib
from nets import inception
from preprocessing import inception_preprocessing


image_size = inception.inception_v1.default_image_size

initialized = False
graph = tf.Graph()

init_fn = None

def predict():
    global initialized
    global init_fn
    with graph.as_default():
        url = 'https://upload.wikimedia.org/wikipedia/commons/7/70/EnglishCockerSpaniel_simon.jpg'
        image_string = urllib.urlopen(url).read()
        image = tf.image.decode_jpeg(image_string, channels=3)
        processed_image = inception_preprocessing.preprocess_image(image, image_size, image_size, is_training=False)
        processed_images  = tf.expand_dims(processed_image, 0)
        with slim.arg_scope(inception.inception_v1_arg_scope()):
            logits, _ = inception.inception_v1(processed_images, num_classes=1001, is_training=False, reuse=initialized)
        probabilities = tf.nn.softmax(logits)
        if not initialized:
            init_fn = slim.assign_from_checkpoint_fn("tmp/checkpoints/inception_v1.ckpt", slim.get_model_variables("InceptionV1"))
        with tf.Session(config=tf.ConfigProto(intra_op_parallelism_threads=8)) as sess:
            init_fn(sess)
            np_probabilities = sess.run(probabilities)
        initialized = True
    return np_probabilities.tolist()

for _ in range(20):
    predict()
```

### Memory Usage monitored with Process Explorer
![memory](https://user-images.githubusercontent.com/20746434/29866915-33ed6506-8d7a-11e7-8dd4-1fc3e1b2f1d7.png)
This is the memory usage when calling the predict function 20 times. As you can see, it keeps increasing.

Am I doing it wrong, or is it a bug?

**Comment by jart** [comment]
@sguada Could you take a look at this potential slim memory leak?

**Comment by ppwwyyxx** [comment]
I think creating ops in loops is the problem.

**Comment by sguada** [comment]
As @ppwwyyxx mentioned you are creating a bigger and bigger graph each time.

Try to create the graph once and use it multiple times, for example

```
def build():
        image_placeholder = tf.placeholder([], tf.string)
        image = tf.image.decode_jpeg(image_placeholder, channels=3)
        processed_image = inception_preprocessing.preprocess_image(image, image_size, image_size, is_training=False)
        processed_images  = tf.expand_dims(processed_image, 0)
        with slim.arg_scope(inception.inception_v1_arg_scope()):
            logits, _ = inception.inception_v1(processed_images, num_classes=1001, is_training=False, reuse=initialized)
        probabilities = tf.nn.softmax(logits)
        return image_placeholder, probabilities

def predict(url, sess, image_placeholder):
        image_string = urllib.urlopen(url).read()
        np_probabilities = sess.run(probabilities, {image_placeholder: image_string})
        return np_probabilities.tolist()

graph = tf.Graph()
with graph.as_default():
  build()
  init_fn = slim.assign_from_checkpoint_fn("tmp/checkpoints/inception_v1.ckpt", slim.get_model_variables("InceptionV1"))
  with tf.Session(config=tf.ConfigProto(intra_op_parallelism_threads=8)) as sess:
    init_fn(sess)

  url = 'https://upload.wikimedia.org/wikipedia/commons/7/70/EnglishCockerSpaniel_simon.jpg'
  for _ in range(20):
      predict(url)
```

**Comment by koelscha** [comment]
Thanks, this solved it. Your code was not running, but the idea was clear, so I could fix it.

Now, if I have requests which use different batch sizes, I would have to reinit the network, right? I tried with resetting the graph, but then I again got a leak...

**Comment by koelscha** [comment]
Just a quick example:

```
import tensorflow as tf
from tensorflow.contrib import slim

from nets import inception

import numpy as np
import random
from time import sleep


initialized = False
batchSize = None

def init(imgs):
    global initialized
    tf.reset_default_graph()
    graph = tf.Graph()
    graph.as_default()
    sess = tf.Session(config=tf.ConfigProto(intra_op_parallelism_threads=8))
    initialized = False
    placeholder, probas = build(imgs)
    init_fn = slim.assign_from_checkpoint_fn("tmp/checkpoints/inception_v1.ckpt", slim.get_model_variables("InceptionV1"))
    init_fn(sess)
    initialized = True
    return sess, placeholder, probas


def build(imgs):
    global batchSize
    batchSize = len(imgs)
    image_placeholder = tf.placeholder(tf.float32, shape=imgs.shape)
    tensorImgs = tf.convert_to_tensor(image_placeholder)
    with slim.arg_scope(inception.inception_v1_arg_scope()):
        logits, _ = inception.inception_v1(tensorImgs, num_classes=1001, is_training=False, reuse=initialized)
    probabilities = tf.nn.softmax(logits)
    return image_placeholder, probabilities


def predict(imgs, image_placeholder, probabilities):
        np_probabilities = sess.run(probabilities, {image_placeholder: imgs})
        return np_probabilities

imgs1 = np.empty((4, 224, 224, 3))
imgs2 = np.empty((2, 224, 224, 3))

sess, placeholder, probas = init(imgs1)
sleep(30)

for _ in range(10):
    print(predict(imgs1, placeholder, probas).argmax())
    sleep(1)

sleep(30)

for _ in range(50):
    if random.choice([0,1]):
        imgs = imgs1
    else:
        imgs = imgs2
    
    if batchSize != len(imgs) or not initialized:
        sess, placeholder, probas = init(imgs)
    print(predict(imgs, placeholder, probas).argmax())
```

Memory usage:
![mem](https://user-images.githubusercontent.com/20746434/30154843-66d3c018-93bb-11e7-9f47-0720c7057f3e.png)

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by sguada** [comment]
You are still building parts of the graph when you change the imgs.

What you need to do is use a placeholder that doesn't specify the batch_size dimension or even the HxW dimensions, it would work for any batch_size and image size, as far all the images in the batch has the same HxW.


```
def init(imgs):
    global initialized
    tf.reset_default_graph()
    graph = tf.Graph()
    graph.as_default()
    sess = tf.Session(config=tf.ConfigProto(intra_op_parallelism_threads=8))
    initialized = False
    image_placeholder = tf.placeholder(tf.float32, shape=[None, None, None, 3])
    probas = build(image_placeholder)
    init_fn = slim.assign_from_checkpoint_fn("tmp/checkpoints/inception_v1.ckpt", slim.get_model_variables("InceptionV1"))
    init_fn(sess)
    initialized = True
    return sess, placeholder, probas


def build(images):
    with slim.arg_scope(inception.inception_v1_arg_scope()):
        logits, _ = inception.inception_v1(images, num_classes=1001, is_training=False, reuse=initialized)
    probabilities = tf.nn.softmax(logits)
    return probabilities
```
